node/src/store/utils.rs

```diff
@@ -1,66 +1,41 @@
-use smallvec::SmallVec;
 
 use crate::executor::yield_now;
-use crate::store::header_ranges::{HeaderRange, HeaderRanges};
+use crate::store::header_ranges::HeaderRange;
 use crate::store::{Result, StoreError};
 
 pub(crate) const VALIDATIONS_PER_YIELD: usize = 4;
 
 /// based on the stored headers and current network head height, calculate range of headers that
 /// should be fetched from the network, starting from the front up to a `limit` of headers
-pub(crate) fn calculate_missing_ranges(
+pub(crate) fn calculate_fetch_range(
     head_height: u64,
     store_headers: &[RangeInclusive<u64>],
     limit: u64,
-) -> HeaderRanges {
-    let range_edges = once(head_height + 1)
-        .chain(
-            store_headers
-                .iter()
-                .flat_map(|r| [*r.start(), *r.end()])
-                .rev(),
-        )
-        .chain(once(0));
-
-    let mut left = limit;
-    HeaderRanges::from_vec(
-        range_edges
-            .chunks(2)
-            .into_iter()
-            .map_while(|mut range| {
-                if left == 0 {
-                    return None;
-                }
-                // ranges_edges is even and we divide into 2 element chunks, this is safe
-                let upper_bound = range.next().unwrap() - 1;
-                let lower_bound = range.next().unwrap();
-                let range_len = upper_bound.checked_sub(lower_bound)?;
-
-                if range_len == 0 {
-                    // empty range
-                    return Some(RangeInclusive::new(1, 0));
-                }
-
-                if left > range_len {
-                    left -= range_len;
-                    Some(lower_bound + 1..=upper_bound)
-                } else {
-                    let truncated_lower_bound = upper_bound - left;
-                    left = 0;
-                    Some(truncated_lower_bound + 1..=upper_bound)
-                }
-            })
-            .filter(|v| !v.is_empty())
-            .collect::<SmallVec<[HeaderRange; 2]>>()
-            .into_iter()
-            .rev()
-            .collect(),
-    )
+) -> HeaderRange {
+    let mut store_headers_iter = store_headers.iter().rev();
+
+    let Some(head_range) = store_headers_iter.next() else {
+        // empty store, just fetch from head
+        return head_height.saturating_sub(limit) + 1..=head_height;
+    };
+
+    if head_range.end() != &head_height {
+        // if we haven't caught up with network head, start from there
+        let fetch_start = u64::max(head_range.end() + 1, head_height.saturating_sub(limit) + 1);
+        return fetch_start..=head_height;
+    }
+
+    // there exists a range contiguous with network head. inspect previous range end
+    let penultimate_range_end = store_headers_iter.next().map(|r| *r.end()).unwrap_or(0);
+
+    let fetch_end = head_range.start().saturating_sub(1);
+    u64::max(
+        penultimate_range_end + 1,
+        fetch_end.saturating_sub(limit) + 1,
+    )..=fetch_end
```

Comment:
I'd split it into two expressions, there is enough math in this function to reason about :D